### DISCUSSION TITLE
Do we need to define machines in both the "darwin" and "home" subtrees?

### DISCUSSION THREAD

**Discussion by seh** [midchildan/dotfiles#94] [Q&A]
Both of the _nix/darwin_ and _nix/home_ directories contain parallel sets of _machines_, _modules_, and _profiles_ directories. Assuming that I'm using macOS as my operating system, it's not clear to me yet whether one must always define configuration for a given machine in both the _nix/darwin/machines_ and the _nix/home/machines_ directories.

The "ci" machine in [the _nix/darwin/machines/ci.nix_ file](https://github.com/midchildan/dotfiles/blob/9d3f1a4b8d8a9203e34e331f46cc18f844829aa3/nix/darwin/machines/ci.nix) appears to configure things for _home-manager_. [The parallel _nix/home/machines/ci.nix_ file](https://github.com/midchildan/dotfiles/blob/1c190d0ac1d87c159b8b7d777f02261ae58a3fc5/nix/home/machines/ci.nix) is similar, but it also [sets the "programs.home-manager.enable" attribute to true](https://github.com/midchildan/dotfiles/blob/1c190d0ac1d87c159b8b7d777f02261ae58a3fc5/nix/home/machines/ci.nix#L3).

I assume that if we're not using _nix-darwin_, then we don't need the machine definition under _nixt/darwin/machines_. If we _are_ using _nix-darwin_, though, do we also need the machine definition under _nix/home/machines_?

**Comment by seh** [comment]
Also, while we're here, why does _nix/darwin/machines/ci.nix_ [start with an arguments set](https://github.com/midchildan/dotfiles/blob/9d3f1a4b8d8a9203e34e331f46cc18f844829aa3/nix/darwin/machines/ci.nix#L1), but _nix/home/machines/ci.nix_ [doesn't need to](https://github.com/midchildan/dotfiles/blob/1c190d0ac1d87c159b8b7d777f02261ae58a3fc5/nix/home/machines/ci.nix#L1)? It appears that they're both modules fed to the "modules" attribute of `darwin.lib.darwinSystem` and `home-manager.lib.homeManagerConfigurations`, respectively.

**Reply by midchildan** [reply]
As mentioned in the [NixOS manual](https://nixos.org/manual/nixos/stable/index.html#sec-configuration-syntax), a module is defined as a function that returns an attribute set.

```nix
{ config, pkgs, ... }:

{
  option = value;
}
```

Alternatively, you can use a plain attribute set instead of a function when you don't need any of the arguments.

```nix
{
  option = value;
}
```

Though the manual is for NixOS, the same thing applies to Home Manager and nix-darwin as well.

**Comment by midchildan** [comment]
Home Manager configuration in `nix/home/machines` and `nix/darwin/machines` are independent of each other. Which one to use depends on how you want to manage Home Manager.

`nix/home/machines` contains configuration for [standalone Home Manager setups](https://nix-community.github.io/home-manager/index.html#sec-install-standalone). This should be used if you want to manage Home Manager separately from nix-darwin. The example configuration in [nix/home/machines/ci.nix](https://github.com/midchildan/dotfiles/blob/1c190d0ac1d87c159b8b7d777f02261ae58a3fc5/nix/home/machines/ci.nix#L3) enables the `programs.home-manager` module to [include the `home-manager` CLI in the user profile](https://github.com/nix-community/home-manager/blob/e3775568184899969115114e72031d226c188934/modules/programs/home-manager.nix#L33).

`nix/darwin/machines` contains configuration for nix-darwin. You can optionally define your Home Manager configuration there, making it possible to build and manage your Home Manager configuration in lockstep with nix-darwin configuration. You can read more about nix-darwin/Home Manager integration [here](https://nix-community.github.io/home-manager/index.html#sec-install-nix-darwin-module). My [example configuration ](https://github.com/midchildan/dotfiles/blob/9d3f1a4b8d8a9203e34e331f46cc18f844829aa3/nix/darwin/machines/ci.nix) for nix-darwin doesn't enable `programs.home-manager` because the `home-manager` CLI shouldn't be used in such a setup.

**Reply by seh** [reply]
Thank you for the thorough reply. Where do we introduce _home-manager_ into the _nix-darwin_ configuration? The documentation to which you linked mentions needing to import `<home-manager/nix-darwin>` in the _nix-darwin_ configuration, but I didn't see that or something similar in [this repository's _nix/darwin/modules/config.nix_ file](https://github.com/midchildan/dotfiles/blob/7cdd097dd01e0678b6ff56487689c78469237722/nix/darwin/modules/config.nix).

**Reply by seh** [reply]
I noticed that I was able to install Nix onto my machine running macOS, but if I then proceeded to install _nix-darwin_, the _darwin-installer_ program would fail, for lack of _home-manager_ being installed. That forced me to back up and install _home-manager_ manually first.

With the way you've set things up with this flake, as I understand it, I can have _nix-darwin_ take care of setting up _home-manager_. Would I need to have installed _nix-darwin_ too, or would this flake take care of it? (It's hard to for me to back up and test this scenario at this point.)

**Reply by midchildan** [reply]
The nix-darwin module provided by Home Manager is [included in `mkDarwin`](https://github.com/midchildan/dotfiles/blob/1c190d0ac1d87c159b8b7d777f02261ae58a3fc5/nix/lib/default.nix#L115), a custom wrapper for [`darwin.lib.darwinSystem`](https://github.com/LnL7/nix-darwin#flakes-experimental). Configuration in `nix/darwin/machines` call `mkDarwin` either directly or [indirectly through `importDarwin`](https://github.com/midchildan/dotfiles/blob/1c190d0ac1d87c159b8b7d777f02261ae58a3fc5/nix/darwin/machines/default.nix#L9).

`mkDarwin` doesn't import `<home-manager/nix-darwin>` though, because that relies on Nix channels and `NIX_PATH`. Those can't be used with flakes. Instead, it imports the same code [exposed through the Home Manager flake](https://nix-community.github.io/home-manager/index.html#sec-flakes-nix-darwin-module).

**Reply by seh** [reply]
Ah, I had missed that `home.darwinModules.default` item there. I found [its definition](https://github.com/nix-community/home-manager/blob/b23c7501f7e0a001486c9a5555a6c53ac7b08e85/flake.nix#L15-L18), which shows _home-manager_ pulling in [its _nix-darwin_ module](https://github.com/nix-community/home-manager/blob/b23c7501f7e0a001486c9a5555a6c53ac7b08e85/nix-darwin/default.nix). It doesn't appear that that installs _nix-darwin_, but rather hooks macOS to activate the _home-manager_ configuration—if I'm reading it correctly. 

I am learning so much studying and adapting your code. It is quite an amazing piece of work.

**Reply by midchildan** [reply]
> if I then proceeded to install nix-darwin, the darwin-installer program would fail, for lack of home-manager being installed

I suppose you imported `<home-manager/nix-darwin>` in your initial nix-darwin configuration? Then you'd need to [add the home-manager channel first](https://nix-community.github.io/home-manager/index.html#sec-install-nix-darwin-module) or Nix won't be able to find `<home-manager/nix-darwin>`.

> With the way you've set things up with this flake, as I understand it, I can have nix-darwin take care of setting up home-manager. Would I need to have installed nix-darwin too, or would this flake take care of it?

The flake takes care of setting up both nix-darwin and Home Manager. But it should be noted that I do some things differently especially with regards to shell configuration.

I take steps to prevent nix-darwin from writing to `/etc/bashrc` and `/etc/zshrc` because [macOS updates tend to overwrite them](https://github.com/NixOS/nix/issues/3616). Instead [I put the required configuration in my shell rc files](https://github.com/midchildan/dotfiles/blob/d106483e133f39d03b8bf30edf42affb0c86fd48/files/.zshenv#L1-L18).

I also [stop nix-darwin from taking over my shell configuration](https://github.com/midchildan/dotfiles/blob/1c190d0ac1d87c159b8b7d777f02261ae58a3fc5/nix/darwin/modules/config.nix#L19-L21) because I prefer to have full control over it. Instead I just use parts of it like [setting environment variables](https://github.com/midchildan/dotfiles/blob/1c190d0ac1d87c159b8b7d777f02261ae58a3fc5/nix/darwin/modules/config.nix#L23-L24) and [enabling shell completion](https://github.com/midchildan/dotfiles/blob/1c190d0ac1d87c159b8b7d777f02261ae58a3fc5/nix/darwin/modules/config.nix#L26-L31). This too requires placing additional configuration in the shell rc files:

https://github.com/midchildan/dotfiles/blob/d106483e133f39d03b8bf30edf42affb0c86fd48/files/.zshenv#L13-L15
https://github.com/midchildan/dotfiles/blob/d106483e133f39d03b8bf30edf42affb0c86fd48/files/.zshrc#L77-L79

> I am learning so much studying and adapting your code. It is quite an amazing piece of work.

Thanks :)

**Reply by seh** [reply]
> I suppose you imported <home-manager/nix-darwin> in your initial nix-darwin configuration?

I took notes, and even so it's still such a blur. Reading them again today, I find the following.

- First, install Nix with the following command:
```shell
curl -L https://nixos.org/nix/install | sh -s -- --daemon
```
(Alternately, `sh <(curl -L https://nixos.org/nix/install)`.)
- Next, I installed Home Manager:  
  (If I tried to skip this step, _nix-darwin_'s _darwin-installer_ program would fail, complaining about not finding what it needs from Home Manager.)
```shell
nix-channel --add https://github.com/rycee/home-manager/archive/master.tar.gz \
            home-manager
 nix-channel --update
```
- Finally, I installed _nix-darwin_:
```shell
nix-build https://github.com/LnL7/nix-darwin/archive/master.tar.gz \
          -A installer 
./result/bin/darwin-installer
```
That failed due to problems with my _ca-certificates.crt_ file, remedied with these commands:
```shell
sudo rm /etc/ssl/certs/ca-certificates.crt 
sudo ln -s /nix/var/nix/profiles/default/etc/ssl/certs/ca-bundle.crt \
           /etc/ssl/certs/ca-certificates.crt
```
I'm hoping you can tell me that some of that shouldn't have been necessary. It felt like too much work, with too many opportunities for failure.

**Reply by midchildan** [reply]
> If I tried to skip this step, nix-darwin's darwin-installer program would fail, complaining about not finding what it needs from Home Manager.

nix-darwin doesn't have any dependency on Home Manager, so such a requirement can only come from configuration. So removing Home Manager stuff from the nix-darwin configuration would've made installing Home Manager unnecessary.

> That failed due to problems with my ca-certificates.crt file, remedied with these commands:

That's odd, but there's just too many reasons certificate verifications could fail so it's hard to tell. Maybe you needed to source `nix-daemon.sh` after installing Nix, or open a new shell to do just that. `/etc/ssl/certs/ca-certificates.crt ` is installed by nix-darwin, so it's also kind of weird that you had to remove it.

**Reply by midchildan** [reply]
Assuming a non-flake setup, only these two should've been necessary if you only wanted nix-darwin:

1. Install Nix
2. Install nix-darwin

**Reply by midchildan** [reply]
The steps I use is written [here](https://github.com/midchildan/dotfiles/blob/main/README.md). Though on macOS, I also need to install Homebrew beforehand because I use Homebrew Cask and [rely on nix-darwin to manage it](https://github.com/midchildan/dotfiles/blob/main/nix/darwin/profiles/apps.nix).

**Reply by seh** [reply]
I studied this command from your documentation for a long while today, and ported the parts necessary to implement it into my repository:
```shell
nix --experimental-features 'nix-command flakes' run '.#os' -- switch --flake '.#'
```
I see how that makes the _darwin-rebuild_ tool—among a few others—available, and uses it to build and activate the top-level default flake target. I have been running _darwin-rebuild build --flake ._ or _darwin-rebuild switch --flake ._ for a few days, given that I already have _darwin-rebuild_ available on my path. If I understand your technique correctly, it doesn't require that _darwin-rebuild_ is available already; instead, it makes it available.

**Reply by midchildan** [reply]
That's correct. The intention was to make bootstrapping easier.
